packages/firestore/src/core/sync_engine.ts

```diff
@@ -721,12 +735,46 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
   }
 
   // PORTING NOTE: Multi-tab only
-  applyPrimaryState(isPrimary: boolean): Promise<void> {
+  async applyPrimaryState(isPrimary: boolean): Promise<void> {
+    const stateUpgraded = this.isPrimary === false && isPrimary;
+
     this.isPrimary = isPrimary;
-    if (this.isPrimary && this.networkAllowed) {
-      return this.remoteStore.enableNetwork();
+    if (this.isPrimary) {
+      if (this.networkAllowed) {
+        await this.remoteStore.enableNetwork();
```

Comment:
FWIW it seems a little odd that we do this even when stateUpgraded == false.  It's tempting to only call enableNetwork() when we become primary and only call disableNetwork() when we lose primary.  So this method might end up:

```
const becamePrimary = this.isPrimary === false && isPrimary;
const lostPrimary = this.isPrimary === true && !isPrimary;
if (becamePrimary) {
  ...
} else if (lostPrimary) {
  ...
}
```

I also wonder if we can/should assert anything about the state of the RemoteStore before enabling the network.  I worry about the state of listenTargets, etc. Are we sure that all of the RemoteStore state is guaranteed to be in a known good state in secondary tabs?  It might be nice to add some asserts or comments about this.

Reply:
I used `becamePrimary` and `lostPrimary` inline. We don't need to do any work when the state doesn't change.

As for brining the RemoteStore into a good state, we need to unlisten all targets when we lose the primary lease (or always call listen/unlisten even in secondary tabs). I haven't added a test for this yet (all existing tests just shut down the client) and will do this work when "throttling" a client becomes possible from the spec tests.